jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs

```diff
@@ -0,0 +1,137 @@
+/*
+ * This software is available under the Apache-2.0 license.
+ * See https://www.apache.org/licenses/LICENSE-2.0.txt for full text.
+ *
+ * Copyright (c) 2024, Gluu, Inc.
+ */
+
+mod error;
+mod openid_config;
+
+pub use error::Error;
+use jsonwebtoken::jwk::JwkSet;
+use jsonwebtoken::DecodingKey;
+use openid_config::*;
+use reqwest::blocking::get;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub struct KeyService {
+    idp_configs: HashMap<Box<str>, OpenIdConfig>, // <issuer (`iss`), OpenIdConfig>
+}
+
+#[allow(unused)]
+impl KeyService {
+    /// initializes a new `KeyService` with the provided OpenID configuration endpoints.
+    ///
+    /// this method fetches the OpenID configuration and the associated keys (JWKS) for each
+    /// endpoint, populating the internal `idp_configs` map. any HTTP errors or parsing
+    /// failures will return a corresponding `Error`.
+    pub fn new(openid_conf_endpoints: Vec<&str>) -> Result<Self, Error> {
+        let mut idp_configs = HashMap::new();
+
+        // fetch IDP configs
+        for endpoint in openid_conf_endpoints {
+            // TODO: use a reusable client when making requests
+            let conf_src: OpenIdConfigSource = get(endpoint)
+                .map_err(Error::HttpError)?
+                .error_for_status()
+                .map_err(Error::HttpError)?
+                .json()
+                .map_err(Error::RequestDeserializationError)?;
+            let (issuer, conf) = OpenIdConfig::from_source(conf_src);
+            idp_configs.insert(issuer, conf);
+        }
+
+        /// retrieves a decoding key based on the provided key ID (`kid`).
+        ///
+        /// this method first attempts to retrieve the key from the local key store. if the key
+        /// is not found, it will refresh the JWKS and try again. if the key is still not found,
+        /// an error of type `KeyNotFound` is returned.
+        for (iss, conf) in &mut idp_configs {
+            let jwks: JwkSet = get(&*conf.jwks_uri)
```

Comment:
and here 
```
client.get(&*conf.jwks_uri)
.send()
```

Reply:
resolved by: [7976b58](https://github.com/JanssenProject/jans/pull/9855/commits/7976b5840cfe5ba7851800dc95325b006f1de8ef)